After two blocks are merged, show cursor on the end of the text field

marecar3 opened the pull request:
Fixes : https://github.com/wordpress-mobile/gutenberg-mobile/issues/699

Support for gutenberg-mobile PR : https://github.com/wordpress-mobile/gutenberg-mobile/pull/756

marecar3 committed:
Control caret position from JS side

marecar3 committed:
Increase condition reusability

marecar3 committed:
Small refactor

marecar3 committed:
Merge branch 'master' into rnmobile/699_Deleting-all-contents-of-block-moves-cursor-to-last-location-in-previous-block

marecar3 committed:
Avoid forcing selection if content is empty

marecar3 committed:
Fixed and operator

marecar3 committed:
Fixed lint errors

marecar3 committed:
Added proper comment

marecar3 committed:
Removed argument type declaration

etoledom reviewed (commented):
Hey @marecar3 - Nice job moving the caret logic to JS 🎉 

As I understand, we are basically moving the caret at the end if the block has been selected but RichText hasn't been set to focus yet. This happens when the block using RichText has been selected programmatically instead of by the user tapping on RichText.

This works well since when the user tap on `RichText` directly, `onFocus` is triggered first, setting isTouched to true and keeping the caret where the user tapped. But when the system set the focus, `componentWillReceiveProps` is called first, and then `onFocus`, moving the caret at the end.

This also work with BottomSheet since when it is dismissed, onFocus is also triggered first.

Please let me know if I'm wrong :)

---

I tried to find other instances where we set focus programmatically where this might be a problem, and I found a small corner case:

When we delete a block, the previous block is automatically selected in a similar way than when merging blocks, so in this case the caret is also moved to the end. I don't think this a big issue and it might be even expected? I'm not sure 🤔 

![deleting_block](https://user-images.githubusercontent.com/9772967/55619490-124a9000-5799-11e9-8777-959c6f4027f5.gif)

I wonder if this logic might also be a problem in future features.

**I did found a small bug in the logic:** 
When we merge into a block that hasn't been selected yet, the caret is still at the beginning. I added a code comment where I think is the problem.

![not_selected](https://user-images.githubusercontent.com/9772967/55619509-1b3b6180-5799-11e9-937d-2bc21b15e927.gif)

---
I miss some comments to clarify what's happening. Maybe to encapsulate the logic inside `componentWillReceiveProps` into a helper function `moveCaretToTheEndIfNeeded()` (or similar explanatory name) would help.

mzorz commented on packages/block-editor/src/components/rich-text/index.native.js:
```diff
@@ -454,6 +470,34 @@ export class RichText extends Component {
+		// This logic will handle the selection when two blocks are merged or when block is split
+		// into two blocks
+		if ( nextTextContent === this.savedContent || nextTextContent.startsWith( this.savedContent ) ) {
+			this.forceSelectionUpdate( this.savedContent.length, this.savedContent.length );
```
good catch here @etoledom , managed to reproduce by starting a draft adding 2 paragraphs then exiting the editor and opening again, then deleting text from the second paragraph

marecar3 replied:
Hey @etoledom, the good catch here. I have fixed the problem but in another way as when I did like you suggest It showed another issue.

marecar3 committed:
Merge branch 'master' into rnmobile/699_Deleting-all-contents-of-block-moves-cursor-to-last-location-in-previous-block

marecar3 committed:
Encapsulate logic into self explanation method
Fixed bug with positioning caret on the start of text when merging two blocks if first has never been selected in the past

marecar3 commented:
Hey, @etoledom thanks for the good review! You have found some good bugs here :)

> When we delete a block, the previous block is automatically selected in a similar way than when merging blocks, so in this case, the caret is also moved to the end. I don't think this a big issue and it might be even expected? I'm not sure 🤔

I have seen similar behavior on Gutenberg web, so I guess it's ok to have it same?

> **I did find a small bug in the logic:**
> When we merge into a block that hasn't been selected yet, the caret is still at the beginning. I added a code comment where I think is the problem.

I think I have fixed that bug now, so you can give it a try.

mkevins commented:
On Android emulator, I'm encountering an issue where I can't merge blocks with backspace.

**To reproduce:**

1. Place caret in the middle of a paragraph block.
2. Press [Enter] to split the block (the caret will be at the start of the new block).
3. Press [Backspace] to merge the blocks.

**Expected result:**

Two paragraph blocks are merged back together.

**Actual result:**

Nothing seems to happen.

**Here is a screenshot:**

<img src="https://user-images.githubusercontent.com/8507675/55721507-e3f1dc80-5a46-11e9-85dc-5e03750124bf.gif" width="360">

etoledom reviewed (approved):
Thank you @marecar3 for the updates!

>I have seen similar behavior on Gutenberg web, so I guess it's ok to have it same?

I agree ✅ 

> I think I have fixed that bug now, so you can give it a try.

Fixed 🎉 

Looks good on iOS!
Let's keep this solution. Hopefully in the future we can have a dispatch action (or similar) to adjust selection from other than just within RichText.

marecar3 committed:
Merge branch 'master' into rnmobile/699_Deleting-all-contents-of-block-moves-cursor-to-last-location-in-previous-block

marecar3 commented:
> On Android emulator, I'm encountering an issue where I can't merge blocks with backspace.
> 
> **To reproduce:**
> 
> 1. Place caret in the middle of a paragraph block.
> 2. Press [Enter] to split the block (the caret will be at the start of the new block).
> 3. Press [Backspace] to merge the blocks.
> 
> **Expected result:**
> 
> Two paragraph blocks are merged back together.
> 
> **Actual result:**
> 
> Nothing seems to happen.
> 
> **Here is a screenshot:**
> 
> <img alt="" width="360" src="https://user-images.githubusercontent.com/8507675/55721507-e3f1dc80-5a46-11e9-85dc-5e03750124bf.gif">

Hey @mkevins nice one!!! :)
Thanks for finding this bug as it was very important and also it didn't work on Android device.
I have fixed it now, so please can you do one more iteration of testing so that we are sure that everything is working? Thanks!

mkevins commented:
I just tested it out and the Android emulator, and it looks like your latest patch fixed the issue I was encountering! 🎉

In the process of testing, I encountered a different issue, but it's not related to this PR. I will open a separate issue to track it.

Thanks @marecar3 !

marecar3 commented:
Thanks @mkevins  and @etoledom !!!

mkevins commented on packages/block-editor/src/components/rich-text/index.native.js:
```diff
@@ -589,7 +642,11 @@ export class RichText extends Component {
 
-		const selection = this.state.needsSelectionUpdate ? { start: this.state.start, end: this.state.end } : null;
+		let selection = null;
+		if ( this.needsSelectionUpdate ) {
```
Just curious, is there a specific rationale for moving the `needsSelectionUpdate` property from the `state` object to the `this` object?

marecar3 replied:
Hey @mkevins, yes, when I tested the code with needsSelectionUpdate being in the state of a component, I saw that we never reset it to false after it was being set to true and we only need needsSelectionUpdate to be true temporary, until we don't send start and end. 

When it was part of the state, if you want to reset it, you would need to make it be false, which means that you would force component to render one more time.

mkevins replied:
That makes sense. Thank you for the explanation!

mkevins commented on packages/block-editor/src/components/rich-text/index.native.js:
```diff
@@ -524,6 +540,43 @@ export class RichText extends Component {
+	}
+
+	forceSelectionUpdate( start, end ) {
+		if ( ! this.needsSelectionUpdate ) {
```
For this method, I'm confused why we don't call `setState` when `needsSelectionUpdate == true`. The name `forceSelectionUpdate`, to me, implies that it will _always_ update the selection.

marecar3 replied:
Hey @mkevins, so it all about resetting it to false, when component renders as I think that it will trigger render one more extra time.

mkevins replied:
I see. I guess the name just threw me off, as the conditional seemed contradictory. As I now understand, it seems this "locks in" the first caller in a race condition.

mkevins reviewed (commented):
Hey @marecar3 :wave: . I just have a few comments I was hoping you could help me with to better understand the changes.

marecar3 commented:
Hey @mkevins, I have left the answers. Let me know if they are clear or if you have some other questions.

mkevins commented:
Thanks @marecar3 for clarifying those points. Your explanation is very helpful!

hypest commented:
I'm a bit confused by the new `savedContent` class variable. We already have `lastContent` and the distinction between the two is not clear. Can we pick another name for the new variable in an attempt to differentiate its use?

marecar3 commented:
Hey @hypest, your suggestion is on the place, we can do it.
Should I open PR with those changes or we can use some active PR for that?
Thanks.

hypest commented:
Hey @marecar3, a dedicated PR would be better to track, review and merge 👍
